addons/mail/static/src/discuss/gif_picker/common/gif_picker.js

```diff
@@ -0,0 +1,268 @@
+    const ref = useRef(refName);
+    const popover = usePopover(GifPicker, {
+        position: "top",
+        popoverClass: "o-fast-popover",
+    });
+    function toggle() {
+        if (popover.isOpen) {
+            popover.close();
+        } else {
+            popover.open(ref.el, options);
+        }
+    }
+    return { toggle };
+}
+
+/**
+ * @typedef {Object} Props
+ * @property {function} onSelected Callback to use when the gif is selected
+ * @property {string} [className]
+ * @property {function} [close]
+ * @extends {Component<Props, Env>}
+ */
+export class GifPicker extends Component {
+    static template = "discuss.GifPicker";
+    static props = ["onSelected", "className?", "close?"];
+
+    async setup() {
+        this.rpc = useService("rpc");
+        this.orm = useService("orm");
+        this.store = useStore();
+        this.userService = useService("user");
+        useAutofocus();
+        useOnBottomScrolled(
+            "scroller",
+            () => {
+                if (!this.state.showCategories) {
+                    this.searchDebounced();
+                }
+            },
+            300
+        );
+        this.next = "";
+        this.state = useState({
+            favorites: [],
+            searchTerm: "",
+            showCategories: true,
+            /** @type {TenorCategory[]} */
+            categories: [],
+            loadingGif: false,
+            loadingError: false,
+            evenGif: {
+                /** @type {TenorGif[]} */
+                gifs: [],
+                /** Size, in pixel, of the column. */
+                columnSize: 0,
+            },
+            oddGif: {
+                /** @type {TenorGif[]} */
+                gifs: [],
+                /** Size, in pixel, of the column. */
+                columnSize: 0,
+            },
+        });
+        this.searchDebounced = useDebounced(this.search, 200);
+        onWillStart(() => {
+            this.loadCategories();
+        });
+        if (!this.store.guest) {
+            onWillStart(() => {
+                this.loadFavorites();
+            });
+        }
+    }
+
+    async loadCategories() {
+        try {
+            const { tags } = await this.rpc(
+                "/discuss/gif/categories",
+                {
+                    country: this.userService.lang.slice(3, 5),
+                    locale: this.userService.lang,
+                },
+                { silent: true }
+            );
+            if (tags) {
+                this.state.categories = tags;
+            }
+        } catch {
+            this.state.loadingError = true;
+        }
+    }
+
+    openCategories() {
+        this.state.showCategories = true;
+        this.state.searchTerm = "";
+        this.clear();
+    }
+
+    closeCategories() {
+        this.state.showCategories = false;
+    }
+
+    async search() {
+        if (!this.state.searchTerm) {
+            return;
+        }
+        this.clear();
```

Comment:
> There is still a problem with this part, I think if there are multiple rpc in progress or something like that, they will also add up results.
I can't reproduce, this is wierd.

> The clear should be done also after the rpc, when starting to add results?

I don't think so, we want to display the loading state as soon as possible, this feel more reactive to the user.

Reply:
>I can't reproduce, this is weird.

I think it could be reproduced if you lower the delay of the `searchDebounced`

Also clearing on search could make the `useOnBottomScrolled` behave strangely, shouldn't it keep appending new gifs as you scroll down instead of resetting?

Reply:
This is because the clear should not happen for each search, but only when the search term changes. I fixed that